File: src/util/Struct.cpp
#include "Struct.hpp"
#include <stdio.h>
#include <stdarg.h>
#include <string.h>

#include "s_load.h"
#include "s_save.h"

/////////////////////////////////////////////

///@cond INTERNAL
Struct::Struct()
{
  clear();
}
Struct::~Struct()
{
  clear();
}
///@endcond
Struct::Struct(const Struct& value)
{
  copy(*this, value);
}
Struct::operator String() const
{
  if((values.size() > 0) && (value == ""))
    const_cast < Struct * > (this)->value = asString();
  return value;
}
Struct::Struct(String value)
{
  clear();
  this->value = value;
}
Struct::Struct(const char *value)
{
  clear();
  this->value = value;
}
Struct::Struct(char value)
{
  clear();
  this->value = value;
}
Struct::operator char() const
{
  String value = (String) * this;
  return value == "" ? '\0' : value[0];
}
Struct::operator bool() const
{
  // Interprets an empty value as false and structured value as true
  if(values.size() > 0)
    return true;
  if(value == "")
    return false;
  // Interprets a numerical value as a boolean
  {
    double v = toDouble(value);
    if(!isnan(v))
      return v != 0;
  }
  // Interprets a string value as a boolean
  {
    std::string s = value;
    std::transform(s.begin(), s.end(), s.begin(), ::tolower);
    return s != "false";
  }
}
Struct::Struct(bool value)
{
  clear();
  this->value = value ? "true" : "false";
}
Struct::operator double() const
{
  return toDouble(value);
}
Struct::Struct(double value)
{
  clear();
  this->value = toName < double > ("%.16g", value);
}
Struct::operator int() const
{
  double v = toDouble(value);
  return isnan(v) ? INT_NAN : (int) rint(v);
}
Struct::operator unsigned int() const
{
  double v = toDouble(value);
  return isnan(v) ? INT_NAN : (unsigned int) rint(v);
}
Struct::Struct(int value)
{
  clear();
  this->value = toName < int > ("%d", value);
}
Struct::Struct(unsigned int value)
{
  clear();
  this->value = toName < unsigned int > ("%u", value);
}
int Struct::toInt(String value)
{
  char *end;
  int r = strtol(value.c_str(), &end, 10);
  return *end == '\0' ? r : INT_NAN;
}
double Struct::toDouble(String value)
{
  char *end;
  double r = strtod(value.c_str(), &end);
  return *end == '\0' ? r : NAN;
}
template < typename T > std::string Struct::toName(const char *format, T value)
{
  static char s[256];
  // - for(int i = 0; i < 256; s[i++] = 0); // used to avoid valgrind erors
  sprintf(s, format, value);
  return s;
}
std::string Struct::toName(int index)
{
  return toName < int > ("%d", index);
}
const Struct Struct::EMPTY;

/////////////////////////////////////////////

void Struct::clear()
{
  for(std::map < std::string, Struct > ::iterator i = values.begin(); i != values.end(); i++)
    i->second.clear();
  values.clear();
  names.clear();
  value = "";
  length = 0;
}
void Struct::copy(Struct& dst, const Struct& src)
{
  dst.clear();
  dst.value = src.value;
  dst.length = src.length;
  for(std::map < std::string, Struct > ::const_iterator i = src.values.begin(); i != src.values.end(); i++)
    copy(dst.values[i->first], i->second);
  for(std::vector < std::string > ::const_iterator i = src.names.begin(); i != src.names.end(); i++)
    dst.names.push_back(*i);
}
void Struct::insert(String name, bool setting)
{
  if(values.find(name) == values.end()) {
    // Manages the names and length
    int l = toInt(name);
    if(l == INT_NAN)
      names.push_back(name);
    else if(l >= length)
      length = l + 1;
    // Resets the corresponding string value
    if(setting)
      this->value = "";
  }
}
void Struct::erase(String name)
{
  if(values.find(name) != values.end()) {
    // Removes the value
    values.erase(name);
    // Manages the names and length
    int l = toInt(name);
    if(l == INT_NAN)
      names.erase(std::remove(names.begin(), names.end(), name), names.end());
    else {
      length = 0;
      for(std::map < std::string, Struct > ::const_iterator i = values.begin(); i != values.end(); i++) {
        int l = toInt(i->first);
        if((l != INT_NAN) && (length < l + 1))
          length = l + 1;
      }
    }
    // Resets the corresponding string value
    this->value = "";
  }
}
void Struct::dump(String header) const
{
  printf("%s value = '%s' count = %d length = %d names = [ ", header.c_str(), value.c_str(), (int) names.size(), length);
  for(std::vector < std::string > ::const_iterator i = names.begin(); i != names.end(); i++)
    printf("'%s' ", (*i).c_str());
  printf("] values = { ");
  for(std::map < std::string, Struct > ::const_iterator i = values.begin(); i != values.end(); i++)
    printf("'%s':«%s»", i->first.c_str(), ((String) i->second).c_str());
  printf("}\n");
}
void Struct::clean()
{
  // Recursively applies on all values
  for(std::map < std::string, Struct > ::iterator i = values.begin(); i != values.end(); i++)
    i->second.clean();
  // Cleans indexed values
  assume(values.size() > 0 || length == 0, "illegal-Struct", "Spurious length %d for an empty or atomic value in '%s'", length, value.c_str());
  for(int i = 0; i < length; i++) {
    String n = toName(i);
    if(values.count(n) > 0) {
      Struct& s = values[n];
      if(s.isEmpty())
        erase(n);
    }
  }
  // Cleans named values
  for(std::vector < std::string > ::iterator i = names.begin(); i != names.end();) {
    String n = *i;
    assume(values.count(n) == 1, "illegal-Struct", "Spurious count %d of named value %s in '%s'\n", values.count(n), n.c_str(), value.c_str());
    Struct& s = values[n];
    if(s.isEmpty())
      erase(n);
    else
      i++;
  }
}
/////////////////////////////////////////////

Struct& Struct::get(String name, Struct value)
{
  if(isEmpty(name))
    set(name, value);
  return get(name);
}
Struct& Struct::get(String name) const
{
  Struct *s = const_cast < Struct * > (this);
  s->insert(name, false);
  return s->values[name];
}
Struct& Struct::set(String name, const Struct& value)
{
  insert(name, true);
  copy(values[name], value);
  return *this;
}
bool Struct::equals(const Struct& value) const
{
  if(isAtomic()) {
    if(value.isAtomic())
      return this->value == value.value;
    else
      return false;
  } else {
    if(value.isAtomic())
      return false;
    else {
      // Calculates the union of both names
      std::set < std::string > names;
      {
        for(std::map < std::string, Struct > ::const_iterator i = values.begin(); i != values.end(); i++)
          names.insert(i->first);
        for(std::map < std::string, Struct > ::const_iterator i = value.values.begin(); i != value.values.end(); i++)
          names.insert(i->first);
      }
      for(std::set < std::string > ::const_iterator i = names.begin(); i != names.end(); i++) {
        if((values.count(*i) == 0) || (value.values.count(*i) == 0))
          return false;
        Struct& s1 = const_cast < Struct & > (*this).values[*i], s2 = const_cast < Struct & > (value).values[*i];
        if(!s1.equals(s2))
          return false;
      }
      return true;
    }
  }
}
void Struct::reset(const char *value, ...)
{
  clear();
  // Performs a sprintf() parsing
  std::string result;
  {
    static const int nchars = 10000;
    static char chars[nchars];
    va_list a;
    va_start(a, value);
    vsnprintf(chars, nchars, value, a);
    va_end(a);
    result = chars;
  }
  // Now reads the String
  if(result != "") {
    // Implements the weak parsing of a JSON structure
    class StructReader {
public:
      void read(Struct& value, String string)
      {
        // Initializes the input buffer
        chars = string.c_str(), index = 0, length = string.length();
        // Clears and set the value
        value.clear();
        read_value(value);
        // Appends the trailer if any
        next_space();
        if(index < length)
          value.add(chars + index);
      }
private:
      // String input buffer, index and length
      const char *chars;
      int index, length;
      std::string word;
      // Reads a value from the string starting at index i and set it
      void read_value(Struct& value)
      {
        next_space();
        switch(chars[index]) {
        case '{':
          read_tuple_value(value);
          break;
        case '[':
          read_list_value(value);
          break;
        default:
          value = read_word();
          break;
        }
      }
      // Reads a word
      String read_word()
      {
        if((chars[index] == '"') || (chars[index] == '\''))
          return read_quoted_word(chars[index]);
        else
          return read_nospace_word();
      }
      // Reads a quoted word
      String read_quoted_word(char quote)
      {
        word = "";
        for(index++; index < length && chars[index] != quote; index++) {
          if((chars[index] == '\\') && (index < length - 1)) {
            index++;
            switch(chars[index]) {
            case '"':
            case '\\':
            case '/':
              word += chars[index];
              break;
            case 'n':
              word += "\n";
              break;
            case 'b':
              word += "\b";
              break;
            case 'r':
              word += "\r";
              break;
            case 't':
              word += "\t";
              break;
            case 'f':
              word += "\f";
              break;
            default:
              word += "\\";
              word += chars[index];
            }
          } else
            word += chars[index];
        }
        if(index < length)
          index++;
        return word;
      }
      // Reads a no-space word
      String read_nospace_word()
      {
        int i0;
        for(i0 = index; index < length && !(isspace(chars[index]) ||
                                            chars[index] == ',' || chars[index] == ';' ||
                                            chars[index] == ':' || chars[index] == '=' ||
                                            chars[index] == '}' || chars[index] == ']'); index++)
        {}
        word = std::string(chars + i0, index - i0);
        return word;
      }
      // Reads a list
      void read_list_value(Struct& value)
      {
        index++;
        while(true) {
          next_space();
          if(index >= length)
            return;
          if(chars[index] == ']') {
            index++;
            return;
          }
          Struct item;
          read_value(item);
          value.set(value.getLength(), item);
          next_space();
          if((index < length) && ((chars[index] == ',') || (chars[index] == ';')))
            index++;
        }
      }
      // Reads a tuple
      void read_tuple_value(Struct& value)
      {
        index++;
        while(true) {
          next_space();
          if(index >= length)
            return;
          if(chars[index] == '}') {
            index++;
            return;
          }
          std::string name = read_word();
          next_space();
          Struct item = true;
          if((index < length) && ((chars[index] == ':') || (chars[index] == '='))) {
            index++;
            read_value(item);
          }
          value.set(name, item);
          next_space();
          if((index < length) && ((chars[index] == ',') || (chars[index] == ';')))
            index++;
        }
      }
      // Shifts until the next non-space char
      void next_space()
      {
        for(; index < length && isspace(chars[index]); index++) {}
      }
    }
    reader;
    reader.read(*this, result);
  }
}
Struct::Struct(int argc, const char *argv[])
{
  clear();
  const char *name = NULL;
  int index = 0;
  bool esc = false;
  for(int i = 1; i < argc; i++) {
    if(strcmp(argv[i], "-") == 0) {
      if(name != NULL) {
        set(name, true);
        name = NULL;
      }
      esc = true;
    } else if((argv[i][0] == '-') && !esc) {
      if(name != NULL)
        set(name, true);
      name = &argv[i][1];
    } else {
      Struct value;
      value.reset(argv[i]);
      if(name == NULL)
        set(index++, value);
      else {
        set(name, value);
        name = NULL;
      }
      esc = false;
    }
  }
  if(name != NULL)
    set(name, true);
}
std::string Struct::asString(String format) const
{
  assume(format == "raw" || format == "plain" || format == "html", "illegal-argument", "in Struct::asString undefined format %s", format.c_str());
  // Implements the string writing of a JSON structure
  class StructWriter {
public:
    // Writes a value in a string, in  compact-raw (mode = 0), 2D-plain-text (mode = 1), html-text (mode = 2).
    void write(std::string& string, const Struct& value, String format = "raw")
    {
      const_cast < Struct & > (value).clean();
      tab = 0, mode = format == "html" ? 2 : format == "plain" ? 1 : 0;
      string = stringHeader();
      write_value(string, value);
      string += stringTrailer();
    }
private:
    // Recursively writes a value
    void write_value(std::string& string, const Struct& value)
    {
      if(value.isAtomic())
        write_word(string, value.value);
      else if(value.getCount() == 0) {
        string += asBeginTag("[");
        for(int i = 0, l = value.getLength() - 1; i <= l; i++) {
          write_value(string, value.get(i));
          string += (i < l ? asNextTag(", ") : asEndTag("]"));
        }
      } else {
        string += asBeginTag("{");
        bool once = true;
        for(Struct::Iterator i(value); i.next();) {
          if(once)
            once = false;
          else
            string += asNextTag(", ");
          write_word(string, i.getName(), true);
          string += asMeta(": ", true);
          write_value(string, value.get(i.getName()));
        }
        for(int i = 0, l = value.getLength() - 1; i <= l; i++) {
          Struct value_i = value.get(i);
          if(!value_i.isEmpty()) {
            if(once)
              once = false;
            else
              string += asNextTag(", ");
            write_word(string, toName(i), true);
            string += asMeta(": ", true);
            write_value(string, value_i);
          }
        }
        string += asEndTag("}");
      }
    }
    void write_word(std::string& string, String value, bool name = false)
    {
      string += asBeginValue("\"", name);
      for(unsigned int i = 0; i < value.length(); i++)
        switch(value[i]) {
        case '"':
        case '\\':
          string += "\\";
          string += value[i];
          break;
        case '\n':
          string += "\\n";
          break;
        case '\b':
          string += "\\b";
          break;
        case '\r':
          string += "\\r";
          break;
        case '\t':
          string += "\\t";
          break;
        case '\f':
          string += "\\f";
          break;
        case '<':
          string += mode == 2 ? "&lt;" : "<";
          break;
        case '&':
          string += mode == 2 ? "&amp;" : "&";
          break;
        default:
          string += value[i];
        }
      string += asEndValue("\"");
    }
    std::string addLn(int t)
    {
      tab += t;
      std::string s;
      if(mode == 1) {
        s += "\n";
        for(int i = 0; i < 2 * tab; i++)
          s += " ";
      }
      return s;
    }
    // HTML and 2D-raw mechanisms
    std::string asBeginTag(String c)
    {
      return mode == 2 ? asMeta(c) + "<div class='struct-block'>" : c + addLn(1);
    }
    std::string asNextTag(String c)
    {
      return mode == 2 ? asMeta(c) + "<br/>" : c + addLn(0);
    }
    std::string asEndTag(String c)
    {
      return mode == 2 ? "</div>" + asMeta(c) : addLn(-1) + c;
    }
    std::string asBeginValue(String c, bool name)
    {
      return mode == 2 ? asMeta(c) + "<span class='" + (name ? "struct-name" : "struct-value") + "'>" : c;
    }
    std::string asEndValue(String c)
    {
      return mode == 2 ? "</span>" + asMeta(c) : c;
    }
    std::string asMeta(String c, bool space = false)
    {
      return mode == 2 ? "<span class='struct-meta-char'>" + (space ? " " + c + " " : c) + "</span>" : mode == 1 && space ? " " + c + " " : c;
    }
    std::string stringHeader()
    {
      return mode == 2 ? "<style>body {background-color:lightgrey}.struct-block{margin-left:20px}.struct-meta-char{color:#330033;font-weight:bold}.struct-name{color:#000066}.struct-value{color:#006600}</style><div class'struct-block'>" : "";
    }
    std::string stringTrailer()
    {
      return mode == 2 ? "</div>" : mode == 1 ? "\n" : "";
    }
    int mode, tab;
  }
  writer;
  std::string string;
  writer.write(string, *this, format);
  return string;
}
void Struct::load(String location)
{
  reset(s_load(location));
}
void Struct::save(String location, String format) const
{
  std::string string = asString(format);
  if(location == "stdout")
    printf("%s", string.c_str());
  else if(location == "stderr")
    fprintf(stderr, "%s", string.c_str());
  else
    s_save(location, string);
}

File: src/util/Struct.js
/** Implements the JavaScript read/write a generic minimal iterative structure in J syntax.
 *
 * @version 0.0.1
 * @copyright <a href='http://www.cecill.info/licences/Licence_CeCILL-C_V1-en.html'>CeCILL-C</a>
 * @author vthierry <thierry.vieville@inria.fr>, Denis Chiron <chiron.denis@gmail.com>
 *
 */
var Struct = {
  /** Parses a string in J syntax and return the related data-structure.
   * @param {string} value The string to parse.
   * @return {object} The parsed data-structure.
   */
  string2data : function(value) {
    var Reader = {
      read : function(value) {
        // Lexical analysis : builds an input of the form
        // [ { tab : line-tabulation, label : label-before-equal, string : line-string } ]
        var input = [];
	for(var index = 0, ll = 0; index < value.length; index++) {
	  var index0, index1, index2, line = true;
	  tab = 0;
	  for(; index < value.length && value[index] != '\n' && this.isspace(value[index]); index++)
	    switch(value[index]) {
	    case '\t':
	      tab += 6;
	      break;
	    case ' ':
	      tab++;
	      break;
	    }
	  // Parses the label before '='
	  for(index0 = index; index < value.length && value[index] != '=' && value[index] != '\n'; index++) ;
	  index1 = value[index] == '=' ? index - 1 : index0;
	  // Parses the value
	  for(; index < value.length && value[index] != '\n' && this.isspace(value[index]); index++) {}
	  if(value[index] == '=') {
	    for(index++; index < value.length && value[index] != '\n' && this.isspace(value[index]); index++) ;
	    for(index2 = index; index < value.length && value[index] != '\n'; index++) ;
	  } else {
	    index1 = index0;
	    for(index2 = index0; index < value.length && value[index] != '\n'; index++) ;
	    // Manages multi-line strings
	    if((ll > 0) && (input[ll - 1]["tab"] <= tab)) {
	      var index3 = index0 - Math.max(0, tab - input[ll - 1]["tab"] - 1);
	      input[ll - 1]["string"] = input[ll - 1]["string"] + "\n" + value.substr(index3, index - index3);
	      line = false;
	    }
	  }
	  // Manages nested data-structure with a string on the label line
	  if(line && (ll > 0) && (input[ll - 1]["tab"] < tab) && (input[ll - 1]["string"] != "")) {
	    input[ll] = { "tab" : tab, "label" : "title", "string" : input[ll - 1]["string"] };
	    input[ll - 1]["string"] = undefined;
	    ll++;
	  }
	  if(line) {
	    input[ll] = { "tab" : tab, "label" : value.substr(index0, index1 - index0), "string" : value.substr(index2, index - index2) };
	    ll++;
	  }
	}
	// -for(var index = 0; index < input.length; index++) console.log(input[index]);
	return this.parse({ input: input, index :0 });
      },
      // Syntax analysis : converts the input to a data-structure
      parse : function(data) {
        var s_value = {}, t_value = [], tab = data.input[data.index]["tab"], size = 0, length = 0;
	for(; data.index < data.input.length && data.input[data.index]["tab"] == tab; data.index++) {
	  var index0 = data.index, item;
	  if((data.index + 1 < data.input.length) && (tab < data.input[data.index + 1]["tab"])) {
	    data.index++;
	    item = this.parse(data);
	  } else {
	    item = this.string2value(data.input[data.index]["string"]);
	  }
	  if(data.input[index0]["label"] == "") {
	    s_value["#" + (length++)] = item;
	    t_value[t_value.length] = item;
	  } else
	    s_value[data.input[index0]["label"]] = item;
	  size++;
	}
	data.index--;
	return size == length ? t_value : s_value;
      },
      string2value : function(string) {
        if(new RegExp("^(true|false)$").test(string))
	  return string == "true";
	else if(new RegExp("^[-+]?[0-9]+$").test(string))
	  return parseInt(string);
	else if(new RegExp("^[-+]?[0-9]*\\.?[0-9]+([eE][-+]?[0-9]+)?$").test(string))
	  return parseFloat(string);
        else
          return string;
      },
      isspace : function(c) {
        return new RegExp("\\s").test(c);
      }
    };
    return Reader.read(value);
  },
  /** Converts a data structure to string in J syntax.
   * @param {object} data The data-structure to render.
   * @param {string} format The string format, either "plain" or "html".
   * @return {string} A string view of the data-structure.
   */
  data2string : function(data, format = "plain") {
    var Writer = {
      write : function(data, format) {
        return (format == "html" ? "<style>body{background-color:lightgrey} .struct-block{margin-left:20px} .struct-meta-char{color:#330033;font-weight:bold} .struct-name{color:#000066} .struct-value{color:#006600} .struct-link{text-decoration: none} .struct-e{font-style: italic} .struct-q{font-family: monospace}</style>\n" : "") +
        this.write_value(data, 0, format) + "\n";
      },
      write_value : function(value, tab, format) {
        var string = "";
	if(!(value instanceof Object))
	  string += this.write_word(String(value), tab, "value", format);
	else if(value instanceof Array)
	  for(var label = 0; label < value.length; label++) {
	    string += this.write_word(format == "html" ? "<div class='struct-block'>" : "\n", tab, "line", format);
	    string += this.write_word("= ", tab, "meta", format);
	    string += this.write_value(value[label], tab + 1, format);
	    string += format == "html" ? "</div>" : ""
	      }
	else {
	  var root = tab == 0, notitle = value["title"] == "" || value["title"] == undefined || root;
	  if(!notitle)
	    string += this.write_word(value["title"], tab, "value", format);
	  for(var label in value)
	    if(notitle || (label != "title")) {
	      string += this.write_word(format == "html" ? "<div class='struct-block'>" : root ? "" : "\n", tab, "line", format);
	      root = false;
	      if(label[0] != '#') {
		string += this.write_word(label, tab, "name", format);
		string += this.write_word(" = ", tab, "meta", format);
	      } else
		string += this.write_word("= ", tab, "meta", format);
	      string += this.write_value(value[label], tab + 1, format);
	      string += format == "html" ? "</div>" : ""
	  }
	}
	return string;
      },
      write_word : function(value, tab, type, format) {
        var string = format != "html" ? "" : type == "meta" ? "<span class='struct-meta-char'>" : type == "name" ? "<span class='struct-name'>" : type == "value" ? "<span class='struct-value'>" : "";
        // Applies the [link] and _span_ transform
        if(format == "html") {
	  var quoted = false;
	  for(var i = 0; i < value.length; i++) {
	    if(value[i] == '"')
	      quoted = !quoted;
	    if(!quoted) {
	      if(value[i] == "[") {
		value = value.substr(0, i) + value.substr(i).replace(new RegExp("\\[([^\\s\\]]+)\\s+([^\\]]*)\\]"), "<a class='struct-link' href='$1'>$2</a>");
		value = value.substr(0, i) + value.substr(i).replace(new RegExp("\\[([^\\s\\]]+)\\]"), "<a class='struct-link' href='$1'>$1</a>");
		i += value.substr(i).search(">");
	      } else if(value[i] == "_") {
		value = value.substr(0, i) + value.substr(i).replace(new RegExp("_([^_])_([^_]+)_"), "<span class='struct-$1'>$2</span>");
		value = value.substr(0, i) + value.substr(i).replace(new RegExp("_([^_]+)_"), "<span class='struct-e'>$1</span>");
		i += value.substr(i).search(">");
	      }
	    }
	  }
	}
	// Applies the \n transform
	for(var i = 0; i < value.length; i++) {
          if(value[i] == '\n') {
	    string += format == "html" && type == "value" ? "</div>\n<div class='struct-block struct-value'>" : "\n";
	    for(var j = 0; j < tab; j++)
	      string += ' ';
	  } else
	    string += value[i];
	}
	string += format != "html" ? "" : type == "meta" || type == "name" || type == "value" ? "</span>" : "";
	return string;
      }
    };
    return Writer.write(data, format);						  
  },
  /** Parses a data structure from a JSON weak-syntax string.
   * @param {string} value The value given as a string, using weak <a href="http://json.org/">JSON</a> syntax.
   * - The weak-syntax, with respect to the strict <a href="https://www.w3schools.com/js/js_json_intro.asp">JSON-syntax</a>allows to:
   *   - use either ':' or '=' between name and value
   *   - use either ',' or ';' as item separator
   *   - use either '"' or "'" as quote
   *   - avoid quotes for strings without space or any meta-char ':=,;]}'
   *   - accept string with '\\n' line separators (replaced by the "\\n" sequence), also manage \\b, \\r, \\t, \\f
   *   - set the value 'true' for name without explicit value
   *   - However: '\\uXXXX' unicode string sequences and the '\/' solidus escaped sequence are not managed (i.e., but simply mirrored in the string value)
   *
   * In other words the input syntax accepts (i) implicit quote <tt>"</tt> when string reduces to one word, (ii) flexible use of comma <tt>,</tt> when optional, (iii) string on several lines, (iv) considering <tt>true</tt> as implicit value, (v) appending as a raw string trailer chars if any.
   * One consequence is that there is no ``syntax error'' all strings map on a JSON structure (i.e., the exact or closest correct JSON structure, the implicit metric being defined by the parsing algorithm). On the reverse the string output is a human readable indented strict JSON syntax allowing to verify that the input was well-formed.
   * @return {object} The parsed data-structure.
   */
  json2data : function(value) {
    var Reader = {
      string : value, 
      index : 0,    
      read : function() {
        var value = this.read_value();
        this.next_space();
        if(this.index < this.string.length)
          return { value : value, trailer : this.string.substr(this.index) };
	return value;
      },
      read_value : function() {
        this.next_space();
        switch(this.string[this.index]) {
        case '{':
          return this.read_tuple_value();
        case '[':
          return this.read_list_value();
        default:
          return this.string2value(this.read_word(true));
        }
      },
      read_tuple_value : function() {
        var value = {};
        this.index++;
        for(var index0 = -1; index0 != this.index;) { index0 = this.index
          this.next_space();
          if(this.index >= this.string.length)
            return value;
          if(this.string[this.index] == '}') {
            this.index++;
            return value;
          }
          var name = this.read_word();
	  if (name == '')
	    return value;
          this.next_space();
          var item = true;
          if((this.index < this.string.length) && ((this.string[this.index] == ':') || (this.string[this.index] == '='))) {
            this.index++;
            item = this.read_value();
          }
          value[name] = item;
          this.next_space();
          if((this.index < this.string.length) && ((this.string[this.index] == ',') || (this.string[this.index] == ';')))
            this.index++;
        }
	console.log({ bug : "read_tuple_value", value : value, string : "«"+this.string.substr(this.index - 100, 100)+"»|«"+this.string.substr(this.index, 10)+"»" });
      },
      read_list_value : function() {
        var value = [];
        this.index++;
        for(var index0 = -1; index0 != this.index;) { index0 = this.index
          this.next_space();
          if(this.index >= this.string.length)
            return value;
          if(this.string[this.index] == ']') {
            this.index++;
            return value;
          }
          value.push(this.read_value());
          this.next_space();
          if((this.index < this.string.length) && ((this.string[this.index] == ',') || (this.string[this.index] == ';')))
            this.index++;
        }
	console.log({ bug : "read_list_value", value : value, string : "«"+this.string.substr(this.index - 100, 100)+"»|«"+this.string.substr(this.index, 10)+"»" });
      },
      read_word : function(line = false) {
        return this.string[this.index] == '"' || this.string[this.index] == '\'' ? this.read_quoted_word(this.string[this.index]) : this.read_nospace_word(line);
      },
      read_quoted_word : function(quote) {
        var word = "";
        for(this.index++; this.index < this.string.length && this.string[this.index] != quote; this.index++) {
          if((this.string[this.index] == '\\') && (this.index < this.string.length - 1)) {
            this.index++;
            switch(this.string[this.index]) {
            case '"':
            case '\\':
            case '/':
              word += this.string[this.index];
              break;
            case 'n':
              word += "\n";
              break;
            case 'b':
              word += "\b";
              break;
            case 'r':
              word += "\r";
              break;
            case 't':
              word += "\t";
              break;
            case 'f':
              word += "\f";
              break;
            default:
              word += "\\";
              word += this.string[this.index];
            }
          } else
            word += this.string[this.index];
        }
        if(this.index < this.string.length)
          this.index++;
        return word;
      },
      read_nospace_word : function(line = false) {
        var i0;
        for(i0 = this.index; this.index < this.string.length && (line ? this.no_endofline(this.string[this.index]) : this.no_space(this.string[this.index])); this.index++) {}
        return this.string.substr(i0, this.index - i0);
      },
      next_space : function() {
        for(; this.index < this.string.length && this.isspace(this.string[this.index]); this.index++);
      },
      string2value : function(string) {
        if(new RegExp("^(true|false)$").test(string))
	  return string == "true";
	else if(new RegExp("^[-+]?[0-9]+$").test(string))
	  return parseInt(string);
	else if(new RegExp("^[-+]?[0-9]*\\.?[0-9]+([eE][-+]?[0-9]+)?$").test(string))
	  return parseFloat(string);
        else
          return string;
      },
      no_space : function(c) {
        return new RegExp("[^\\s,;:=}\\]]").test(c);
      },
      no_endofline : function(c) {
        return new RegExp("[^\\n,;:=}\\]]").test(c);
      },
      isspace : function(c) {
        return new RegExp("\\s").test(c);
      }
    };
    return Reader.read();
  },
  /** Converts a string from a format to another format.
   * @param {string} what The conversion to perform:
   * - 'j2json' : Converts a string from J syntax to JSON syntax.
   * - 'json2j' : Converts a string from JSON syntax to J syntax.
   * - 'j2html' : Converts a string in J syntax to HTML for display.
   * - 'j2j' : Reformats a J syntax string to a nomalized form.
   * - 'wjson2json' : Reformats a weak JSON syntax string to a normalized form.
   * - 'json2json' : Reformats a JSON syntax string to a normalized form.
   * @param {string} value The string to parse in the input format.
   * @return {string} The parsed string in the output format.
   */
  convert : function(what, value) {
    switch(what) {
    case 'j2json':
      return JSON.stringify(Struct.string2data(value), null, 2);
    case 'json2j':
      return Struct.data2string(JSON.parse(value));
    case 'j2html':
      return Struct.data2string(Struct.string2data(value), "html");
    case 'j2j':
      return Struct.data2string(Struct.string2data(value));
    case 'wjson2json':
      return JSON.stringify(Struct.json2data(value), null, 2);
    case 'json2json':
      return JSON.stringify(JSON.parse(value), null, 2);
    default:
      return value;
    }
  }
};

global.Struct = Struct;
